### PR TITLE
SDI-360 Update Nomis from incentive creation domain event

### PR DESCRIPTION
**Pull request by steverendell**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderBooking.kt [hunk 1]
```diff
@@ -113,6 +113,15 @@ data class OffenderBooking(
   @OneToMany(mappedBy = "id.offenderBooking", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
   val incentives: MutableList<Incentive> = mutableListOf(),
 ) {
+  fun getNextSequence(): Long {
+    val i = incentives.stream().max(Comparator.comparing { (a, _) -> a.sequence })
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderBooking.kt:117]: You could have used Kotlin semantics here `incentives.map(it.sequence).maxOrNull()?. let {it + 1}   ?: 1`  or something like that

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/IncentivesService.kt [hunk 2]
```diff
@@ -1,27 +1,64 @@
...
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun createIncentive(bookingId: Long, dto: CreateIncentiveRequest): CreateIncentiveResponse {
+    val offenderBooking = offenderBookingRepository.findById(bookingId)
+      .orElseThrow(NotFoundException(bookingId.toString()))
+
+    val mappedIncentive = mapIncentiveModel(dto, offenderBooking)
+    val incentive = incentiveRepository.save(mappedIncentive)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/IncentivesService.kt:45]: not sure why we need to save here? Can't we just add to booking and let the transaction commit?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/IncentivesService.kt [hunk 3]
```diff
@@ -67,4 +104,25 @@ class IncentivesService(
...
+  private fun mapIncentiveModel(dto: CreateIncentiveRequest, offenderBooking: OffenderBooking): Incentive {
+
+    val sequence = offenderBooking.getNextSequence()
+
+    val iepLevel = iepLevelRepository.findById(IEPLevel.pk(dto.iepLevel))
+      .orElseThrow(BadDataException("Invalid IEP type from incentives: ${dto.iepLevel}"))
+
+    val location = agencyLocationRepository.findById(dto.agencyId)
+      .orElseThrow(BadDataException("Prison with id=${dto.agencyId} does not exist"))
+
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/IncentivesService.kt:117]: I think you also need to check the incentive level exist for the prison as well (sorry can't remember table)
In fact, if you just check IEP level/location exists then the 2 above calls are redundant I think?

**Review by andymarke — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderBooking.kt (hunk 1 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderBooking.kt:117]: Im glad you mentioned that ! , i couldnt work out what the 'native' way forward was with this

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/IncentivesService.kt (hunk 2 above)
Comment (steverendell) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/service/IncentivesService.kt:45]: good point

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**